### DISCUSSION TITLE
Best way to do graphics

### DISCUSSION THREAD

**Discussion by AlquistArj** [QuestPDF/QuestPDF#580] [General]
I need to create a graphics heavy report which is a kind of colour coded heat map. I'm uncertain how best to create the graphic. Would I be best creating a .png and inserting it onto the page, or should I try drawing each coloured "pixel" as a filled shape directly on the .pdf page?

If I go with the .png option I'm basically doing all the page rendering as a graphic and just using QuestPDF to convert the .png to a .pdf.

However, I can imagine drawing individual pixels on the .pdf page could be large / slow to render.

What would you advise?

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out.

For pixel-dense graphics like heat maps, rendering as a raster image (PNG) and embedding it via `.Image()` is the recommended approach — drawing thousands of individual shapes would significantly increase file size and slow down rendering. You can also control image quality and DPI via `.WithRasterDpi()` and `.WithCompressionQuality()` to balance visual fidelity and file size.

For vector-based graphics, QuestPDF now also supports native [SVG](https://www.questpdf.com/api-reference/svg.html) rendering and [Complex Graphics](https://www.questpdf.com/api-reference/complex-graphics.html).

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
